### ISSUE TITLE
Scavenge not cleaning up any data based on max age

### ISSUE THREAD

**Issue by keukenkastjeXYZ** [EventStore/EventStore#2625]
**Describe the bug**
We use event-store via the following docker container with docker-compose:
https://hub.docker.com/r/eventstore/eventstore (we use tag: release-5.0.8)
And in a Java Spring Boot application with the following library:
https://github.com/msemys/esjc (we use version 2.2.0)

With this setup we connect successfully to the event-store with an application and we set the max age of the (single) stream we use to 15 minutes (900 seconds). Every 30 minutes we also run a scheduler inside the application, that requests a scavenge successfully to the event store. We would suspect that with this scavenge older events would be cleaned due to the max age we set, but that does not happen, all events are kept in place.

**To Reproduce**
-, I am not sure how you could reproduce this. You could try to run event store in a docker compose, connect via the ESJC client library and set the max age on the stream like:
   
  ```
 @Bean
    public EventStore eventStore(final EventStoreConnectionProperties connectionProperties,
                                 final EventStoreCleanupProperties cleanupProperties) {
        final EventStore eventStore = EventStoreBuilder.newBuilder()
                .singleNodeAddress(connectionProperties.getHost(), connectionProperties.getPort())
                .userCredentials(connectionProperties.getUsername(), connectionProperties.getPassword())
                .maxReconnections(connectionProperties.getRetries())
                .build();

        eventStore.setStreamMetadata(streamName, ExpectedVersion.ANY,
                StreamMetadata.newBuilder()
                        .maxAge(cleanupProperties.getMaxAge())
                        .build());

        return eventStore;
    }
```
Even though, this more seems like a problem in event store itself to me. The max age is properly set. Events get inserted successfully and the scavenge also runs fine. Only problem is that the scavenge does not clean anything up.

**Expected behavior**
A scavenge clears up all events older then 15 minutes, as bound to the set max-age on the stream.

**Actual behavior**
Almost no space gets cleared/becomes free and the database grows and grows until the whole disk is full. Events older then 15 minutes stay in the event store. 

**Config/Logs/Screenshots**
  ```[00001,53,13:17:06.621] Scavenging "#51-51 (chunk-000051.000000)" traversed 427 including 427.  
[00001,53,13:17:06.621] Scavenging of chunks:  
"#51-51 (chunk-000051.000000)"  
completed in 00:00:18.3931807.  
Old chunks' versions are kept as they are smaller.  
Old chunk total size: 267091968, scavenged chunk size: 267096064.  
Scavenged chunk removed.  
[00001,53,13:17:06.621] File "46289c10-c09e-49b8-83ba-300940f0f996.scavenge.tmp" has been marked for delete and will be deleted in TryDestructFileStreams.  
[00001,53,13:17:06.621] SCAVENGING: started to scavenge chunks: "#52-52 (chunk-000052.000000)" 52 => 52 (13958643712 => 14227079168)  
[00001,53,13:17:06.621] Resulting temp chunk file: "fe08c0b9-0f64-471a-b2e1-c38ce5558ed7.scavenge.tmp". 
```

**Comment by gregoryyoung** [comment]
This is telling you why they were not removed.

The result of the scavenge would end up with a larger data file. 

When a scavenge occurs it requires writing a map in order handle positions being changed. If the number of events to be removed is very small the scavenge can result in **more** disk space being used. As such the system by default will prefer the smaller file.

There is an option around this listed on https://eventstore.com/docs/server/command-line-arguments/index.html which is AlwaysKeepScavengedChunks. This will override this behaviour and keep the scavenged version even if the scavenged version results in a larger file.

**Comment by keukenkastjeXYZ** [comment]
@gregoryyoung thank you for the quick response! The problem for us regarding the scavenging is that we for instance have 54 chunks of events. That the last chunk does not get picked up into the scavenging process is no problem for us. Our event store consists of one stream with a time to live of 900 seconds (15 minutes). This also gets returned when you request the metadata of the stream in the eventstore. Then after more then 15 minutes a scavenge is successfully requested and starts running. What we would expect to happen is for all the events that have been in the stream for more then 15 minutes to be scavenged, while that does not happen. Instead no events get deleted.

Now the question remains: does the time to live on a stream actually soft delete the events? Or is there another way of (soft) deleting events after 15 minutes, so that they can be picked up by the scavenging process? 

Thanks in advance!

**Comment by gregoryyoung** [comment]
How are you measuring this?

Remember when scavenge occurs only the data files which are **not the current** for TFchunks are scavenged. 

In the case ^^^ the log message is telling you that it didnt scavenge them because it would lose space by doing so. You can force this to occur if you want but will lose space by doing so...

The TTL says "these are eligible for scavenge" by default. Whether or not to scavenge is a secondary decision.

**Comment by keukenkastjeXYZ** [comment]
@gregoryyoung Thanks for your reply. I am not sure if my problem is clear. I understand how the scavenging process works, that it will only scavenge the non-current chunks and that there is an option to keep the bigger scavenge chunk file over the smaller existing chunk file.

However, to me it seems that the events in the stream do not get marked as "soft deleted" or "ready for being scavenged", which is something I would expect to happen after the maxage from the stream has passed for an event. So for me the problem is not that the scavenge does not loop over the correct chunks or keeps/deletes the scavenges chunks, but rather that I would expect 40 out of for instance 45 chunks to be cleaned from events as the maxage from the stream has passed.

Is above assumption correct? Or do we need to use another way to mark events as "soft deleted" to get them ready for the scavenging process?

**Comment by keukenkastjeXYZ** [comment]
@gregoryyoung just to be sure we also tried to set the proposed property AlwaysKeepScavengedChunks, but as expected that did also not help. Do you have any updates on the questions from the previous question?

**Comment by keukenkastjeXYZ** [comment]
@gregoryyoung after all we have been able to resolve the issue. In the end the problem was the one custom JavaScript projection that we added to the event store. It consisted out of two streams, one for events and one for "snapshots" of events. This snapshot event stream ran on every new event that was added to the store, while it should have ran every 5/6 minutes (this was our fault). A snapshot in our case meant a snapshot of all the events that were in the store. This meant that on every new event, a snapshot of all the events in the store was made. Because we had a throughput of for instance 5-6 events per second, this meant the event store was growing rapidly in size, creating a snowball effect. This also caused the CPU, memory and disk to overflow, get slow and in the end crash. 

We now have created a "sidecar" application that creates the event snapshots as expected every 5/6 minutes with a Spring Boot scheduler implementation. After all the scavenging works as expected with the maxage on the stream.
